### PR TITLE
Clean up and added test cases for `data_objects.data_containers`

### PR DESCRIPTION
**Pull request by git-abhishek**
## PR Summary
1. Removed dead code
2. Fixed few issues which were identified by adding more test cases
3. Increased coverage
 
## PR Checklist
- [X] Code passes flake8 checker
- [ ] New features are documented, with docstrings and narrative docs **NA**
- [X] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by git-abhishek** [comment]
Unable to add @[Colin](https://github.com/colinmarc) as a reviewer. If someone can add him, that would be great!

File: yt/data_objects/tests/test_data_containers.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    ds = fake_particle_ds()
+    sp = ds.sphere(ds.domain_center, 0.25)
+    sp.write_out(filename)
+
+    with open(filename, "r") as file:
+        file_row_1 = file.readline()
+        file_row_2 = file.readline()
+        file_row_2 = np.array(file_row_2.split('\t'), dtype=np.float64)
+    sorted_keys = sorted(sp.field_data.keys())
+    _keys = [str(k) for k in sorted_keys]
```
Comment (colinmarc) [inline yt/data_objects/tests/test_data_containers.py]: I don't see the point of signifying these as private with `_`.

File: yt/data_objects/tests/test_data_containers.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    loaded_sphere = obj["my_sphere_1"][1]
+    assert_array_equal(loaded_sphere.center, sp.center)
+    assert_equal(loaded_sphere.radius, sp.radius)
+    obj.close()
+
+    sp.save_object("my_sphere_2")
+
+def test_to_dataframe():
+    try:
+        import pandas as pd
```
Comment (colinmarc) [inline yt/data_objects/tests/test_data_containers.py]: Does `import pandas as _` make flake8 happy?

File: yt/data_objects/tests/test_data_containers.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    proj.to_frb((1.0, 'unitary'), 64)
+
+def test_extract_isocontours():
+    # Test isocontour properties for AMRGridData
+    ds = fake_amr_ds(fields=["density", "cell_mass"], particles=16**3)
+    dd = ds.all_data()
+    rho = dd.quantities["WeightedAverageQuantity"]("density",
+                                                   weight="cell_mass")
+    dd.extract_isocontours("density", rho, "triangles.obj", True)
+    dd.calculate_isocontour_flux("density", rho, "x", "y", "z",
```
Comment (colinmarc) [inline yt/data_objects/tests/test_data_containers.py]: This formatting choice seems odd (the complete line is shorter than the lines above). Are you using an auto-formatter like yapf or eyeballing it?

File: yt/data_objects/tests/test_data_containers.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    _data = [sp.field_data[k][0] for k in sorted_keys]
+
+    assert_equal(_keys, file_row_1)
+    assert_array_equal(_data, file_row_2)
+
+def test_save_object():
+    ds = fake_particle_ds()
+    sp = ds.sphere(ds.domain_center, 0.25)
+    sp.save_object("my_sphere_1", filename="test_save_obj")
+    obj = shelve.open("test_save_obj", protocol=-1)
```
Comment (cphyc) [inline yt/data_objects/tests/test_data_containers.py]: This could be wrapped in a `with` statement to prevent the file handler to being left opened if the assertion bellow fail.

File: yt/data_objects/construction_data_containers.py [hunk 5]
```diff
@@ -922,7 +922,8 @@ def __init__(self, *args, **kwargs):
         self._final_start_index = self.global_startindex
 
     def _setup_data_source(self, level_state = None):
-        if level_state is None: return
+        if level_state is None:
+            return super(YTSmoothedCoveringGrid, self)._setup_data_source()
```
Comment (ngoldbaum) [inline yt/data_objects/construction_data_containers.py]: I'd format this as:

```
if level_state is None:
    super(YTSmoothedCoveringGrid, self)._setup_data_source()
    return
```

`_setup_data_source` doesn't return anything, and this way of calling it makes it clearer that that's the case.

File: yt/data_objects/data_containers.py [hunk 6]
```diff
@@ -1673,8 +1664,8 @@ def to_frb(self, width, resolution, center=None, height=None,
                     "Currently we only support images centered at R=0. " +
                     "We plan to generalize this in the near future")
             from yt.visualization.fixed_resolution import CylindricalFixedResolutionBuffer
-            if iterable(width):
-                radius = max(width)
+            if isinstance(width, tuple):
+               radius = width[0]
```
Comment (ngoldbaum) [inline yt/data_objects/data_containers.py]: I don't think this change is correct - the original code is correct. We're checking for `iterable(width)` because width might be a list or an ndarray or some other kind of iterable. It's easier to just use the `iterable` function to test for this rather than using `isinstance` and insisting on only accepting a certain number of types.

Also it's a behavior change to pick `width[0]` instead of `max(width)`. Can you explain why you did that? Choosing `max(width)` is arbitrary, but since we already made the decision it's best to not change our minds without a good reason.

File: yt/data_objects/tests/test_data_containers.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+from yt.testing import assert_equal, fake_random_ds, fake_amr_ds, fake_particle_ds
+from yt.utilities.exceptions import YTFieldNotFound
+
+
+def test_yt_data_container():
+    # Test if ds could be None
+    with assert_raises(RuntimeError) as err:
+        YTDataContainer(None, None)
+    desired = 'Error: ds must be set either through class type or parameter' \
+              ' to the constructor'
```
Comment (ngoldbaum) [inline yt/data_objects/tests/test_data_containers.py]: style nit: If you put the string inside parentheses then you don't need to use the line continuation

File: yt/data_objects/tests/test_data_containers.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    assert_equal(str(err.exception), desired)
+
+    # Test if field_data key exists
+    ds = fake_random_ds(5)
+    proj = ds.proj("density", 0, data_source=ds.all_data())
+    assert_equal('px' in proj.keys(), True)
+    assert_equal('pz' in proj.keys(), False)
+
+    # Delete the key and check if exits
+    proj.__delitem__('px')
```
Comment (ngoldbaum) [inline yt/data_objects/tests/test_data_containers.py]: Rather than calling `__delitem__` directly I'd format this (and similarly below) as `del proj['px']`

File: yt/data_objects/tests/test_data_containers.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    assert_equal('density' in proj.keys(), False)
+
+    # Delete a non-existent field
+    with assert_raises(YTFieldNotFound) as ex:
+        proj.__delitem__('p_mass')
+    desired = "Could not find field '('stream', 'p_mass')' in UniformGridData."
+    assert_equal(str(ex.exception), desired)
+
+    # Test the convert method
+    assert_equal(proj.convert('HydroMethod'), -1)
```
Comment (ngoldbaum) [inline yt/data_objects/tests/test_data_containers.py]: This is actually a relic of the old pre yt-3.0 API that should have been removed a long time ago. Can you remove this test and the function calling it?

File: yt/data_objects/tests/test_data_containers.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+def test_write_out():
+    filename = "sphere.txt"
+    ds = fake_particle_ds()
+    sp = ds.sphere(ds.domain_center, 0.25)
+    sp.write_out(filename)
```
Comment (ngoldbaum) [inline yt/data_objects/tests/test_data_containers.py]: Now that you've added a test for this function (awesome!) would you mind also adding a docstring in `yt/data_containers.py`?

File: yt/data_objects/tests/test_data_containers.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    assert_array_equal(_data, file_row_2)
+
+def test_save_object():
+    ds = fake_particle_ds()
+    sp = ds.sphere(ds.domain_center, 0.25)
+    sp.save_object("my_sphere_1", filename="test_save_obj")
+    obj = shelve.open("test_save_obj", protocol=-1)
+    loaded_sphere = obj["my_sphere_1"][1]
+    assert_array_equal(loaded_sphere.center, sp.center)
+    assert_equal(loaded_sphere.radius, sp.radius)
```
Comment (ngoldbaum) [inline yt/data_objects/tests/test_data_containers.py]: Maybe also test field access?

File: yt/data_objects/tests/test_data_containers.py (hunk 2 above)
Comment (ngoldbaum) [reply yt/data_objects/tests/test_data_containers.py]: Rather than wrapping the test logic in a try/except, please use `yt.testing.requires_module`, which is a decorator you can apply to a test function to tell nose that the test function requires some optional module (in this case `pandas`).

You probably also need to install pandas on travis and appveyor to get this test to actually execute.

File: yt/data_objects/tests/test_data_containers.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+def test_std():
+    ds = fake_random_ds(3)
+    ds.all_data().std('density', weight="velocity_z")
+
+def test_to_frb():
+    ds = fake_amr_ds(fields=["density", "cell_mass"], geometry="cylindrical",
+                     particles=16**3)
+    proj = ds.proj("density", weight_field="cell_mass", axis=1,
+                   data_source=ds.all_data())
+    proj.to_frb((1.0, 'unitary'), 64)
```
Comment (ngoldbaum) [inline yt/data_objects/tests/test_data_containers.py]: Can you test that the `FixedResolutionBuffer` object you get back is correct? (e.g. it produces images with a resolution of `(64, 64)` and that the image covers the full domain? For the latter you can just inspect the state of the `FixedResolutionBuffer` object (e.g. by checking `frb.bounds`) rather than trying to make sure the image is correct.

File: yt/data_objects/tests/test_data_containers.py (hunk 3 above)
Comment (ngoldbaum) [reply yt/data_objects/tests/test_data_containers.py]: I'd format these two lines like:

```
dd.calculate_isocontour_flux(
    "density", rho, "x", "y", "z", "dx")
```

File: yt/data_objects/tests/test_data_containers.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+def test_extract_isocontours():
+    # Test isocontour properties for AMRGridData
+    ds = fake_amr_ds(fields=["density", "cell_mass"], particles=16**3)
+    dd = ds.all_data()
+    rho = dd.quantities["WeightedAverageQuantity"]("density",
+                                                   weight="cell_mass")
```
Comment (ngoldbaum) [inline yt/data_objects/tests/test_data_containers.py]: I'd format these lines like:

```
q = dd.quantities["WeightedAverageQuantity"]
rho = q("density", weight="cell_mass")
```

Often when you find yourself hitting a line length limit you can get around it by defining an intermediate variable. This has the side benefit of often making the code more understandable.

File: yt/data_objects/tests/test_data_containers.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+def test_extract_connected_sets():
+
+    ds = fake_random_ds(16, nprocs=8, particles=16 ** 3)
+    data_source = ds.disk([0.5, 0.5, 0.5], [0., 0., 1.], (8, 'kpc'), (1, 'kpc'))
+    field = ("gas", "density")
+    min_val, max_val = data_source[field].min() / 2, data_source[field].max() / 2
+
+    data_source.extract_connected_sets(field, 3, min_val, max_val,
+                                              log_space=True, cumulative=True)
```
Comment (ngoldbaum) [inline yt/data_objects/tests/test_data_containers.py]: Can you test that the data returned by this function are correct? Also the formatting right here is a little odd, this line should be aligned with `field`, not `3`.

File: yt/data_objects/data_containers.py (hunk 6 above)
Comment (git-abhishek) [reply yt/data_objects/data_containers.py]: I changed it as I was getting error for the test case:
```
def test_to_frb():
    ds = fake_amr_ds(fields=["density", "cell_mass"], geometry="cylindrical",
                     particles=16**3)
    proj = ds.proj("density", weight_field="cell_mass", axis=1,
                   data_source=ds.all_data())
    proj.to_frb((1.0, 'unitary'), 64)
```
Error:
```
line 1668, in to_frb
    radius = max(width)
TypeError: '>' not supported between instances of 'str' and 'float'
```

Comment (ngoldbaum) [reply yt/data_objects/data_containers.py]: Width is allowed to be a tuple of tuples, like this:

```
((1.0, 'unitary'), (0.7, 'unitary'))
```

this allows constructing images where the width doesn't equal the height.

For the cylindrical FRB the width really represents a radius, and what happens right now is it assumes if you're passing a tuple you're passing two floats (which probably needs to be relaxed) and it assumes the radius you *really* mean is the maximum of the two floats you passed in.

That's not great, so maybe it's ok to change this. Rather than what you've done though I'd suggest if we're passed a tuple, to check if it's a valid `(length, unit)` tuple, and if not raise an error. That is, for a cylindrical dataset, it only makes sense to pass in one width to inrepret as the radius, so if someone passes in more than one, we should raise an error.

Basically the code here:

https://github.com/yt-project/yt/blob/cc299b07a5923b28c9105bd898e231412f25d3cc/yt/data_objects/data_containers.py#L1691-L1696

But in the if statement raise an error instead of interpreting the tuple as a width and height.

File: yt/data_objects/tests/test_data_containers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply yt/data_objects/tests/test_data_containers.py]: Don't delete `__delitem__`, that's still needed. I was saying that *in the test* don't invoke `__delitem__` directly, instead use the `del` statement to invoke it indirectly.

You should delete `convert` since it's a relic function that doesn't do what it says in its own docstring.

File: yt/data_objects/tests/test_data_containers.py (hunk 10 above)
Comment (git-abhishek) [reply yt/data_objects/tests/test_data_containers.py]: done

File: .travis.yml [hunk 15]
```diff
@@ -23,6 +23,7 @@ env:
     FLAKE8=flake8
     CODECOV=codecov
     COVERAGE=coverage
+    PANDAS=pandas<0.21
```
Comment (ngoldbaum) [inline .travis.yml]: is there some problem with the latest version of pandas?

File: yt/data_objects/data_containers.py [hunk 16]
```diff
@@ -215,6 +192,10 @@ def _set_center(self, center):
                 self.center = self.ds.find_max(("gas", "density"))[1]
             elif center.startswith("max_"):
                 self.center = self.ds.find_max(center[4:])[1]
+            elif center.lower() == "min":
+                self.center = self.ds.find_min(("gas", "density"))[1]
+            elif center.startswith("min_"):
+                self.center = self.ds.find_min(center[4:])[1]
```
Comment (ngoldbaum) [inline yt/data_objects/data_containers.py]: good catch!

File: yt/data_objects/tests/test_data_containers.py [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+    sp.save_object("my_sphere_1", filename="test_save_obj")
+    obj = shelve.open("test_save_obj", protocol=-1)
+    loaded_sphere = obj["my_sphere_1"][1]
+    obj.close()
+    assert_array_equal(loaded_sphere.center, sp.center)
+    assert_equal(loaded_sphere.radius, sp.radius)
+    for k in loaded_sphere._key_fields:
+        assert_array_equal(loaded_sphere[k], sp[k])
+
+    # Object is saved but retrieval is not working
```
Comment (ngoldbaum) [inline yt/data_objects/tests/test_data_containers.py]: can you open an issue to track this?

File: .travis.yml (hunk 15 above)
Comment (git-abhishek) [reply .travis.yml]: I was getting ` Double requirement given: numpy==1.12.1 ... (already in numpy==1.9.3 )` and build was failing at [Travis](https://travis-ci.org/yt-project/yt/jobs/392083344#L652).
Solution: https://github.com/pandas-dev/pandas/issues/20723

Comment (ngoldbaum) [reply .travis.yml]: Ah, so it's because pandas dropped support for `python3.4` - can you only add this version restriction for the python3.4 builder?

To be honest, we should probably drop testing for python3.4, it's going EOL - I'll make a note to bring that up at the team meeting tomorrow.

Comment (git-abhishek) [reply .travis.yml]: Our Travis lint stage is executed for python3.4, do you want me to update that too with python3.6?

Comment (ngoldbaum) [reply .travis.yml]: Yup.

File: yt/data_objects/tests/test_data_containers.py (hunk 17 above)
Comment (git-abhishek) [reply yt/data_objects/tests/test_data_containers.py]: done

File: yt/data_objects/tests/test_data_containers.py (hunk 4 above)
Comment (cphyc) [reply yt/data_objects/tests/test_data_containers.py]: OK, fair enough!

File: yt/data_objects/data_containers.py [hunk 18]
```diff
@@ -427,20 +401,73 @@ def _parameter_iterate(self, seq):
...
+            msg = ("Field type %s of the supplied field %s is inconsistent"
+                   " with field type '%s'." %
+                   ([f[0] for f in f_type], [f[1] for f in f_type], req_ftype))
+            raise YTException(msg)
+
         for field in field_order: self[field]
-        field_order += [field for field in fields if field not in field_order]
-        fid = open(filename,"w")
-        fid.write("\t".join(["#"] + field_order + ["\n"]))
+        fid = open(filename, "w")
```
Comment (cphyc) [inline yt/data_objects/data_containers.py]: I'd be happier with a `with` statement there, to prevent the file handler from being left opened.

Comment (git-abhishek) [reply yt/data_objects/data_containers.py]: done

**Comment by ngoldbaum** [comment]
@yt-fido test this please
